compiler/rustc_typeck/src/check/pat.rs

```diff
@@ -649,39 +649,38 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
-    fn borrow_pat_suggestion(
-        &self,
-        err: &mut Diagnostic,
-        pat: &Pat<'_>,
-        inner: &Pat<'_>,
-        expected: Ty<'tcx>,
-    ) {
+    fn borrow_pat_suggestion(&self, err: &mut Diagnostic, pat: &Pat<'_>, inner: &Pat<'_>) {
         let tcx = self.tcx;
-        if let PatKind::Binding(..) = inner.kind {
+        if let PatKind::Ref(_, mutbl) = pat.kind
+        && let PatKind::Binding(_, _, binding, ..) = inner.kind {
             let binding_parent_id = tcx.hir().get_parent_node(pat.hir_id);
             let binding_parent = tcx.hir().get(binding_parent_id);
-            debug!("inner {:?} pat {:?} parent {:?}", inner, pat, binding_parent);
+            debug!(?inner, ?pat, ?binding_parent);
+
+            let mutability = match mutbl {
+                ast::Mutability::Mut => "mut",
+                ast::Mutability::Not => "",
+            };
+
             match binding_parent {
-                hir::Node::Param(hir::Param { span, .. })
-                    if let Ok(snippet) = tcx.sess.source_map().span_to_snippet(inner.span) =>
-                {
-                    err.span_suggestion(
-                        *span,
-                        &format!("did you mean `{snippet}`"),
-                        format!(" &{expected}"),
-                        Applicability::MachineApplicable,
+                hir::Node::Param(hir::Param { ty_span, .. }) if binding.span.hi() <= ty_span.lo() => {
```

Comment:
This is meant to check for closures without explicit type... is there a better way to do this?

Reply:
that is, trying to filter out `|_|`? 

Maybe we can reverse the polarity of the check -- in the case of `|_|` is there something special about `ty_span` (is it empty? is it equal to `binding.span`) we can instead use in the `if` statement on the match arm?

Reply:
This tries to filter out things like `|x|` where the type is inferred. And yes, in this case `ty_span` is equal to `binding.span` (or is it? anyway it points to something useless/weird and overlaps with the binding).

I would actually expect that if there is no type `ty_span` is dummy, but it is not the case.

Reply:
Can you change this to `ty_span == binding_span` then? (Or I think there is some special span equality method)